builtin/k8s/platform.go

```diff
@@ -296,318 +294,392 @@ func (p *Platform) resourceDeploymentStatus(
-			TimeoutSeconds:      timeoutSeconds,
-			FailureThreshold:    failureThreshold,
-		}
+	// Setup our port configuration
+	if p.config.ServicePort == 0 && len(p.config.Ports) == 0 {
+		// nothing defined, set up the defaults
+		p.config.Ports = append(p.config.Ports, &Port{Port: DefaultServicePort, Name: "http"})
+	} else if p.config.ServicePort > 0 && len(p.config.Ports) == 0 {
+		// old ServicePort var is used, so set it up in our Ports map to be used
+		p.config.Ports = append(p.config.Ports, &Port{Port: p.config.ServicePort, Name: "http"})
+	} else if p.config.ServicePort > 0 && len(p.config.Ports) > 0 {
+		// both defined, this is an error
+		return fmt.Errorf("cannot define both 'service_port' and 'ports'. Use" +
+			" 'ports' for configuring multiple container ports")
+	}
 
-		deployment.Spec.Template.Spec.Containers[0].ReadinessProbe = &corev1.Probe{
-			Handler: corev1.Handler{
-				HTTPGet: &corev1.HTTPGetAction{
-					Path: p.config.ProbePath,
-					Port: intstr.FromInt(defaultPort),
+	envVars := make(map[string]string)
+	// Add deploy config environment to container env vars
+	for k, v := range p.config.StaticEnvVars {
+		envVars[k] = v
+	}
+	for k, v := range deployConfig.Env() {
+		envVars[k] = v
+	}
+
+	// Check autoscaling
+	if p.config.AutoscaleConfig != nil && p.config.CPU == nil {
+		ui.Output("For autoscaling in Kubernetes to work, a deployment must specify "+
+			"cpu resource limits and requests. Otherwise the metrics-server will not properly be able "+
+			"to scale your deployment.", terminal.WithWarningStyle())
+	}
+
+	// Create scratch space volumes
+	var volumes []corev1.Volume
+	for idx := range p.config.ScratchSpace {
+		scratchName := fmt.Sprintf("scratch-%d", idx)
+		volumes = append(volumes,
+			corev1.Volume{
+				Name: scratchName,
+				VolumeSource: corev1.VolumeSource{
+					EmptyDir: &corev1.EmptyDirVolumeSource{},
 				},
 			},
-			InitialDelaySeconds: initialDelaySeconds,
-			TimeoutSeconds:      timeoutSeconds,
-			FailureThreshold:    failureThreshold,
-		}
+		)
 	}
 
-	if len(p.config.ScratchSpace) > 0 {
-		for idx, scratchSpaceLocation := range p.config.ScratchSpace {
-			scratchName := fmt.Sprintf("scratch-%d", idx)
-			deployment.Spec.Template.Spec.Volumes = append(
-				deployment.Spec.Template.Spec.Volumes,
-				corev1.Volume{
-					Name: scratchName,
-					VolumeSource: corev1.VolumeSource{
-						EmptyDir: &corev1.EmptyDirVolumeSource{},
-					},
-				},
-			)
+	// Configure containers
+	var command *[]string
+	var args *[]string
+	if p.config.Pod != nil && p.config.Pod.Container != nil {
+		command = p.config.Pod.Container.Command
+		args = p.config.Pod.Container.Args
+	}
+
+	appContainer, err := configureK8sContainer(
+		src.App,
+		fmt.Sprintf("%s:%s", img.Image, img.Tag),
+		p.config.Ports,
+		envVars,
+		p.config.Probe,
+		p.config.ProbePath,
+		p.config.CPU,
+		p.config.Memory,
+		p.config.Resources,
+		command,
+		args,
+		p.config.ScratchSpace,
+		volumes,
+		log,
```

Comment:
If we update `SidecarContainer` struct to just be `Container`, we can initialize a container struct before this and just pass it through rather than having a lot of params.